src/plugins/data/common/search/aggs/agg_configs.ts

```diff
@@ -196,57 +218,110 @@ export class AggConfigs {
+        const prevDsl = dslLvlCursor[prevConfig.id];
+
+        // advance the cursor and nest under the previous agg, or
+        // put it on the same level if the previous agg doesn't accept
+        // sub aggs
+        dslLvlCursor = prevDsl?.aggs || dslLvlCursor;
+      }
 
-        const dsl = config.type.hasNoDslParams
-          ? config.toDsl(this)
-          : (dslLvlCursor[config.id] = config.toDsl(this));
-        let subAggs: any;
+      if (hasMultipleTimeShifts) {
+        dslLvlCursor = this.insertTimeShiftSplit(config, timeShifts, dslLvlCursor);
+      }
 
-        parseParentAggs(dslLvlCursor, dsl);
+      if (config.type.hasNoDsl) {
+        return;
+      }
 
-        if (config.type.type === AggGroupNames.Buckets && i < list.length - 1) {
-          // buckets that are not the last item in the list accept sub-aggs
-          subAggs = dsl.aggs || (dsl.aggs = {});
-        }
+      const dsl = config.type.hasNoDslParams
+        ? config.toDsl(this)
+        : (dslLvlCursor[config.id] = config.toDsl(this));
+      let subAggs: any;
 
-        if (subAggs) {
-          _.each(subAggs, (agg) => {
-            parseParentAggs(subAggs, agg);
-          });
-        }
-        if (subAggs && nestedMetrics) {
-          nestedMetrics.forEach((agg: any) => {
-            subAggs[agg.config.id] = agg.dsl;
-            // if a nested metric agg has parent aggs, we have to add them to every level of the tree
-            // to make sure "bucket_path" references in the nested metric agg itself are still working
-            if (agg.dsl.parentAggs) {
-              Object.entries(agg.dsl.parentAggs).forEach(([parentAggId, parentAgg]) => {
-                subAggs[parentAggId] = parentAgg;
-              });
-            }
-          });
-        }
-      });
+      parseParentAggs(dslLvlCursor, dsl);
+
+      if (
+        config.type.type === AggGroupNames.Buckets &&
+        (i < aggsWithDsl - 1 || timeSplitIndex > i)
+      ) {
+        // buckets that are not the last item in the list of dsl producing aggs or have a time split coming up accept sub-aggs
+        subAggs = dsl.aggs || (dsl.aggs = {});
+      }
+
+      if (subAggs) {
+        _.each(subAggs, (agg) => {
+          parseParentAggs(subAggs, agg);
+        });
+      }
+      if (subAggs && nestedMetrics) {
+        nestedMetrics.forEach((agg: any) => {
+          subAggs[agg.config.id] = agg.dsl;
+          // if a nested metric agg has parent aggs, we have to add them to every level of the tree
+          // to make sure "bucket_path" references in the nested metric agg itself are still working
+          if (agg.dsl.parentAggs) {
+            Object.entries(agg.dsl.parentAggs).forEach(([parentAggId, parentAgg]) => {
+              subAggs[parentAggId] = parentAgg;
+            });
+          }
+        });
+      }
+    });
 
     removeParentAggs(dslTopLvl);
     return dslTopLvl;
   }
 
+  private insertTimeShiftSplit(
+    config: AggConfig,
```

Comment:
can we please add one of these (either):
- comments to this file explaining how we are building request and how we are transforming response (pretty much what you had in that gist we used for discussion)
- or put the same in .md file in this folder
- or put it in the PR description

i think it will be useful in the future